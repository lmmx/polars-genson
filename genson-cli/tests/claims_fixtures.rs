use assert_cmd::Command;
use insta::{assert_snapshot, with_settings};
use std::fs;

/// Check if the current output matches the verified/blessed version
fn is_output_approved(snapshot_name: &str, output: &str) -> bool {
    let module_file = file!();
    let module_stem = std::path::Path::new(module_file)
        .file_stem()
        .unwrap()
        .to_string_lossy();
    let verified_path = format!("tests/verified/{}__{}.snap", module_stem, snapshot_name);

    if let Ok(verified_content) = fs::read_to_string(&verified_path) {
        if let Some(header_end) = verified_content.find("\n---\n") {
            let verified_output = &verified_content[header_end + 5..];
            return verified_output.trim() == output.trim();
        }
    }
    false
}

/// Run genson-cli with claims fixture from disk
fn run_genson_claims_fixture_from_disk(fixture_path: &str, name: &str, extra_args: &[&str]) {
    let mut cmd = Command::cargo_bin("genson-cli").unwrap();
    let mut args = vec![
        "--map-threshold",
        "0",
        "--unify-maps",
        "--wrap-root",
        "claims",
    ];
    // Add --ndjson flag if the fixture is a .jsonl file
    if fixture_path.ends_with(".jsonl") {
        args.push("--ndjson");
    }
    args.extend_from_slice(extra_args);
    args.push(fixture_path);
    let args_for_metadata = args.clone();
    cmd.args(args);

    let assert_output = cmd.assert().success();
    let output = assert_output.get_output();
    let stdout_str = String::from_utf8(output.stdout.clone()).unwrap();

    // Let stderr be visible if there's any debug output
    if !output.stderr.is_empty() {
        let stderr_str = String::from_utf8_lossy(&output.stderr);
        eprintln!("stderr from {}: {}", name, stderr_str);
    }

    let approved = is_output_approved(name, &stdout_str);

    with_settings!({
        info => &serde_json::json!({
            "approved": approved,
            "args": args_for_metadata[..args_for_metadata.len()-1], // exclude file path
            "fixture": fixture_path
        })
    }, {
        assert_snapshot!(name, stdout_str);
    });
}

#[test]
fn test_claims_fixture_l1_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l1_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l1_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l2_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l2_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l2_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l3_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l3_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l3_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__normalize",
        &["--normalise"],
    );
}

// Following are numbered from the full x1818 fixture

#[test]
fn test_claims_fixture_x1818_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__normalize",
        &["--normalise"],
    );
}
