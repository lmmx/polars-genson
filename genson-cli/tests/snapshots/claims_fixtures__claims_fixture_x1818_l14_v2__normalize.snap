---
source: genson-cli/tests/claims_fixtures.rs
expression: stdout_str
info:
  approved: false
  args:
    - "--map-threshold"
    - "0"
    - "--unify-maps"
    - "--wrap-root"
    - claims
    - "--normalise"
    - "--force-type"
    - "mainsnak:record"
    - "--no-unify"
    - qualifiers
  fixture: tests/data/claims/x1818_L14_PENULTIMATE.json
---
[
  {
    "claims": {
      "P487": [
        {
          "mainsnak": {
            "property": "P487",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "üêï"
            },
            "datatype": "string",
            "property-labels": {
              "en": "Unicode character",
              "de": "Unicodezeichen"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P805": [
              {
                "property": "P805",
                "datavalue": {
                  "datavalue__string": null,
                  "id": "Q87577372",
                  "labels": {
                    "egl": "üêï",
                    "ase": "üêï"
                  }
                },
                "datatype": "wikibase-item",
                "property-labels": {
                  "en": "statement is subject of",
                  "ca": "afirmaci√≥ detallada a"
                }
              }
            ]
          },
          "references": null
        },
        {
          "mainsnak": {
            "property": "P487",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "üê∂"
            },
            "datatype": "string",
            "property-labels": {
              "en": "Unicode character",
              "de": "Unicodezeichen"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P805": [
              {
                "property": "P805",
                "datavalue": {
                  "datavalue__string": null,
                  "id": "Q87577442",
                  "labels": {
                    "egl": "üê∂",
                    "ase": "üê∂"
                  }
                },
                "datatype": "wikibase-item",
                "property-labels": {
                  "en": "statement is subject of",
                  "ca": "afirmaci√≥ detallada a"
                }
              }
            ]
          },
          "references": null
        }
      ],
      "P1245": [
        {
          "mainsnak": {
            "property": "P1245",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "966"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "OmegaWiki Defined Meaning",
              "de": "OmegaWiki-Bedeutungskennung"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P1036": [
        {
          "mainsnak": {
            "property": "P1036",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "636.7"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "Dewey Decimal Classification",
              "de": "Dewey-Dezimalklassifikation"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": [
            {
              "P248": [
                {
                  "property": "P248",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q36578",
                    "labels": {
                      "de": "Gemeinsame Normdatei",
                      "en": "Integrated Authority File"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "fr": "affirm√© dans",
                    "ca": "afirmat a"
                  }
                }
              ],
              "P227": [
                {
                  "property": "P227",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": null,
                    "labels": null,
                    "datavalue__string": "4026181-5"
                  },
                  "datatype": "external-id",
                  "property-labels": {
                    "en": "GND ID",
                    "de": "GND-Kennung"
                  }
                }
              ],
              "P407": [
                {
                  "property": "P407",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q188",
                    "labels": {
                      "tg": "–∑–∞–±–æ–Ω–∏ –æ–ª–º–æ–Ω”£",
                      "ay": "Aliman aru"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "es": "idioma de la obra o del nombre",
                    "pt": "l√≠ngua da obra ou do nome"
                  }
                }
              ],
              "P813": [
                {
                  "property": "P813",
                  "datavalue": {
                    "time": "+2021-08-29T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "id": null,
                    "labels": null,
                    "datavalue__string": null
                  },
                  "datatype": "time",
                  "property-labels": {
                    "en": "retrieved",
                    "en-ca": "retrieved"
                  }
                }
              ]
            }
          ]
        },
        {
          "mainsnak": {
            "property": "P1036",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "599.772"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "Dewey Decimal Classification",
              "de": "Dewey-Dezimalklassifikation"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": [
            {
              "P248": [
                {
                  "property": "P248",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q36578",
                    "labels": {
                      "de": "Gemeinsame Normdatei",
                      "en": "Integrated Authority File"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "fr": "affirm√© dans",
                    "ca": "afirmat a"
                  }
                }
              ],
              "P227": [
                {
                  "property": "P227",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": null,
                    "labels": null,
                    "datavalue__string": "4026181-5"
                  },
                  "datatype": "external-id",
                  "property-labels": {
                    "en": "GND ID",
                    "de": "GND-Kennung"
                  }
                }
              ],
              "P407": [
                {
                  "property": "P407",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q188",
                    "labels": {
                      "tg": "–∑–∞–±–æ–Ω–∏ –æ–ª–º–æ–Ω”£",
                      "ay": "Aliman aru"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "es": "idioma de la obra o del nombre",
                    "pt": "l√≠ngua da obra ou do nome"
                  }
                }
              ],
              "P813": [
                {
                  "property": "P813",
                  "datavalue": {
                    "time": "+2021-08-29T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "id": null,
                    "labels": null,
                    "datavalue__string": null
                  },
                  "datatype": "time",
                  "property-labels": {
                    "en": "retrieved",
                    "en-ca": "retrieved"
                  }
                }
              ]
            }
          ]
        }
      ],
      "P373": [
        {
          "mainsnak": {
            "property": "P373",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "Dogs"
            },
            "datatype": "string",
            "property-labels": {
              "or": "‡¨ï‡¨Æ‡¨®‡≠ç‡¨∏ ‡¨∂‡≠ç‡¨∞‡≠á‡¨£‡≠Ä",
              "oc": "categoria Commons"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": [
            {
              "P143": [
                {
                  "property": "P143",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q8449",
                    "labels": {
                      "fr": "Wikip√©dia en espagnol",
                      "en": "Spanish Wikipedia"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "nl": "ge√Ømporteerd uit Wikimedia-project",
                    "de": "importiert aus Wikimedia-Projekt"
                  }
                }
              ]
            }
          ]
        }
      ],
      "P508": [
        {
          "mainsnak": {
            "property": "P508",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "285"
            },
            "datatype": "external-id",
            "property-labels": {
              "it": "identificativo Thesaurus BNCF",
              "en": "BNCF Thesaurus ID"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P1810": [
              {
                "property": "P1810",
                "datavalue": {
                  "datavalue__string": "Cani",
                  "id": null,
                  "labels": null
                },
                "datatype": "string",
                "property-labels": {
                  "en": "subject named as",
                  "ru": "–Ω–∞–∑–≤–∞–Ω –∫–∞–∫"
                }
              }
            ]
          },
          "references": null
        }
      ],
      "P910": [
        {
          "mainsnak": {
            "property": "P910",
            "datavalue": {
              "id": "Q6830323",
              "labels": {
                "el": "ŒöŒ±œÑŒ∑Œ≥ŒøœÅŒØŒ±:Œ£Œ∫œçŒªŒøœÇ",
                "en": "Category:Dogs"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "topic's main category",
              "pt": "categoria principal do t√≥pico"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P646": [
        {
          "mainsnak": {
            "property": "P646",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "/m/0bt9lr"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "Freebase ID",
              "fa": "ÿ¥ŸÜÿßÿ≥ŸáŸî ŸÅÿ±€å‚Äåÿ®€åÿ≥"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": [
            {
              "P248": [
                {
                  "property": "P248",
                  "datavalue": {
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "id": "Q15241312",
                    "labels": {
                      "en": "Freebase Data Dumps",
                      "de": "Freebase-Datendump"
                    },
                    "datavalue__string": null
                  },
                  "datatype": "wikibase-item",
                  "property-labels": {
                    "fr": "affirm√© dans",
                    "ca": "afirmat a"
                  }
                }
              ],
              "P577": [
                {
                  "property": "P577",
                  "datavalue": {
                    "time": "+2013-10-28T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "id": null,
                    "labels": null,
                    "datavalue__string": null
                  },
                  "datatype": "time",
                  "property-labels": {
                    "fr": "date de publication",
                    "it": "data di pubblicazione"
                  }
                }
              ]
            }
          ]
        }
      ],
      "P227": [
        {
          "mainsnak": {
            "property": "P227",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "4026181-5"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "GND ID",
              "de": "GND-Kennung"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P1810": [
              {
                "property": "P1810",
                "datavalue": {
                  "datavalue__string": "Hund",
                  "id": null,
                  "labels": null
                },
                "datatype": "string",
                "property-labels": {
                  "en": "subject named as",
                  "ru": "–Ω–∞–∑–≤–∞–Ω –∫–∞–∫"
                }
              }
            ],
            "P4970": [
              {
                "property": "P4970",
                "datavalue": {
                  "datavalue__string": "Canis familiaris",
                  "id": null,
                  "labels": null
                },
                "datatype": "string",
                "property-labels": {
                  "en": "alternative name",
                  "pt": "nomes alternativos"
                }
              },
              {
                "property": "P4970",
                "datavalue": {
                  "datavalue__string": "Canis lupus familiaris",
                  "id": null,
                  "labels": null
                },
                "datatype": "string",
                "property-labels": {
                  "en": "alternative name",
                  "pt": "nomes alternativos"
                }
              },
              {
                "property": "P4970",
                "datavalue": {
                  "datavalue__string": "Haushund",
                  "id": null,
                  "labels": null
                },
                "datatype": "string",
                "property-labels": {
                  "en": "alternative name",
                  "pt": "nomes alternativos"
                }
              }
            ]
          },
          "references": null
        }
      ],
      "P417": [
        {
          "mainsnak": {
            "property": "P417",
            "datavalue": {
              "id": "Q152457",
              "labels": {
                "lb": "Rochus vu Montpellier",
                "vec": "San Roco"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "patron saint",
              "es": "santo patr√≥n"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        },
        {
          "mainsnak": {
            "property": "P417",
            "datavalue": {
              "id": "Q159834",
              "labels": {
                "pl": "Hubert z Li√®ge",
                "he": "◊î◊ï◊ë◊®◊ò◊ï◊°"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "patron saint",
              "es": "santo patr√≥n"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P1051": [
        {
          "mainsnak": {
            "property": "P1051",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "886"
            },
            "datatype": "external-id",
            "property-labels": {
              "hu": "PSH-azonos√≠t√≥",
              "fr": "identifiant Polytematick√Ω strukturovan√Ω hesl√°≈ô"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P279": [
        {
          "mainsnak": {
            "property": "P279",
            "datavalue": {
              "id": "Q57814795",
              "labels": {
                "en": "domesticated mammal",
                "hu": "h√°zias√≠tott eml≈ës"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "subclass of",
              "ca": "subclasse de"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        },
        {
          "mainsnak": {
            "property": "P279",
            "datavalue": {
              "id": "Q39201",
              "labels": {
                "nan": "Thi√≥ng-buÃçt",
                "zh": "ÂÆ†Áâ©"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "subclass of",
              "ca": "subclasse de"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P935": [
        {
          "mainsnak": {
            "property": "P935",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "Dog"
            },
            "datatype": "string",
            "property-labels": {
              "en": "Commons gallery",
              "pl": "galeria Commons"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P2167": [
        {
          "mainsnak": {
            "property": "P2167",
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "10101502"
            },
            "datatype": "external-id",
            "property-labels": {
              "en": "UNSPSC Code",
              "ru": "UNSPSC"
            }
          },
          "rank": "normal",
          "qualifiers": null,
          "references": null
        }
      ],
      "P1343": [
        {
          "mainsnak": {
            "property": "P1343",
            "datavalue": {
              "id": "Q2041543",
              "labels": {
                "sk": "Ott≈Øv slovn√≠k nauƒçn√Ω",
                "ru": "–ù–∞—É—á–Ω–∞—è —ç–Ω—Ü–∏–∫–ª–æ–ø–µ–¥–∏—è –û—Ç—Ç–æ"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "described by source",
              "ru": "–æ–ø–∏—Å—ã–≤–∞–µ—Ç—Å—è –≤ –∏—Å—Ç–æ—á–Ω–∏–∫–∞—Ö"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P805": [
              {
                "property": "P805",
                "datavalue": {
                  "datavalue__string": null,
                  "id": "Q23864606",
                  "labels": {
                    "ru": "–û—Ç—Ç–æ / —Å–æ–±–∞–∫–∞",
                    "cs": "Pes"
                  }
                },
                "datatype": "wikibase-item",
                "property-labels": {
                  "en": "statement is subject of",
                  "ca": "afirmaci√≥ detallada a"
                }
              }
            ]
          },
          "references": null
        },
        {
          "mainsnak": {
            "property": "P1343",
            "datavalue": {
              "id": "Q602358",
              "labels": {
                "it": "Dizionario Enciclopedico Brockhaus ed Efron",
                "az": "Brokhauz v…ô Efronun ensiklopedik l√ºƒü…ôti"
              },
              "datavalue__string": null
            },
            "datatype": "wikibase-item",
            "property-labels": {
              "en": "described by source",
              "ru": "–æ–ø–∏—Å—ã–≤–∞–µ—Ç—Å—è –≤ –∏—Å—Ç–æ—á–Ω–∏–∫–∞—Ö"
            }
          },
          "rank": "normal",
          "qualifiers": {
            "P805": [
              {
                "property": "P805",
                "datavalue": {
                  "datavalue__string": null,
                  "id": "Q24484874",
                  "labels": {
                    "labels__string": "–≠–°–ë–ï / –°–æ–±–∞–∫–∏"
                  }
                },
                "datatype": "wikibase-item",
                "property-labels": {
                  "en": "statement is subject of",
                  "ca": "afirmaci√≥ detallada a"
                }
              }
            ]
          },
          "references": null
        }
      ]
    }
  }
]
